### ISSUE TITLE
User must verify their account on www.facebook.com (405) via spectrum2

### ISSUE THREAD

**Issue by truenicoco** [dequis/purple-facebook#474]
This is similar to #282 #278 and #274, but since they are closed, I figured I should open a new ticket.

### Errors thrown by pidgin

Not pidgin, but by gajim talking to spectrum2 2.0.12-1 through prosody 0.11.3:

>User must verify their account on www.facebook.com (405)

### Message shown by the facebook website (particularly pay attention to mentions of "unfamiliar location" or "suspicious activity")

Nothing at all during my latest attempts. In the past, I had to indeed verify through different means (friends, email, check comments) and had to change passwords a few times.

### Is it all the same internet connection?

Yes. prosody+spectrum run on a small server I rent. For a while I have only been connecting to facebook via a SOCKS proxy on this same server, which I can verify works because the location of my active connections are the server's location, and facebook started recommending events near this location instead of my real one. Both are in France. Since I can browse messenger.com and facebook.com via my SOCKS proxy, this is not an IP ban, right?

### Are you using any proxies / VPNs?

See above.

### Were you using app passwords? Did the situation improve after using app passwords?

I tried both. When I first set up my xmpp server + spectrum2 transports, everything worked fine for about a year, despite me logging to the web interface from several different locations. A few times I had to verify my account and chage passwords, but it kept on working after that. Then since a few months, it was either not possible to log in via spectrum2, or I got kicked out after a few days. Last attempt that worked I used app password, it worked for a few days only. Now my login attempts do not even result in a warning on the web interface.

### Did it happen once or did it happen on every login attempt?

Every attempt now.

### Was it enough to change the password?

Sometimes yes, sometimes no. I have been trying every few weeks for a few months now.

### Have you used this plugin before? If so, when was the last time it worked?

About a week ago, for a few days only.

### Have you used any other third party facebook clients like miranda?

Not miranda, but 2 android unofficial apps: frost and event sync for facebook. I thought they were responsible about the first warnings, but I cannot be sure. Anyway, I stopped using them for several months and only connected through the web via a SOCKS5 proxy.

**Comment by zyphlar** [comment]
Having the same issue via Bitlbee ( https://github.com/bitlbee/bitlbee-facebook/issues/172 and https://github.com/bitlbee/bitlbee-facebook/issues/128 ) not sure if it's related to this package or not. (bitlbee reports plugin version `facebook 20190203~c76b36b~331`, the january 2019 version is even worse with SSL errors)

**Comment by thorgrin** [comment]
I ran into this issue tonight after restarting bitlbee, the facebook logs me out with 405 and locks my account. I tried connecting with browser from the same IP address the bitlbee is running, but there is no change. After trying to log in from bitlbee, the account gets locked.

My version is 1.1.2-1 on ubuntu-xenial.

**Comment by truenicoco** [comment]
It has been a while, so I tried again, but trying to login either via Finch or spectrum2 still locks my account, even using an app password.
Connecting via the messenger or facebook web interface from the same IP just works, so I guess it really purple-facebook that is doing something that facebook doesn't like. I wish I had the skills to debug this…

**Comment by truenicoco** [comment]
A little update: for the past two weeks, it was working, even though I didn't change anything. But since yesterday, it stopped working again, with new error messages. First "Unable to connect to b-api.facebook.com: SSL Connection Failed", and now it can't connect with "Internal Server Error, please reconnect.". I don't know if it is worth opening a new issue for this.

**Comment by dhwalker** [comment]
I just installed purple-facebook 20190113-0.9.6-0250907-166 under Linux Mint 19 (Ubuntu 18.04), and I'm also getting the "SSL Connection Failed" error.  Has this been resolved, or is there a workaround?

**Comment by truenicoco** [comment]
Every once in a while, I check whether it would work again on my setup. Yesterday it did! For less than 24hours before receiving a:

`The action attempted has been deemed abusive or is otherwise disallowed`

before getting to the usual

`User must verify their account on www.facebook.com (405)`,

which seems to confirm that this purple-facebook (or spectrum2?) is doing something that Facebook considers "abusive". Don't know what though.

**Comment by zyphlar** [comment]
I also was getting a ton of "abusive" disconnects from my laptop running
pidgin yesterday. I think their abuse algorithm was acting funky.

**Comment by truenicoco** [comment]
Unfortunately it wasn't just yesterday in my case. Any attempt at reconnecting ends up in locking my account (again). Oh well, it was nice while it was working but I'll just continue checking messenger.com everyonce in a while.

**Comment by truenicoco** [comment]
FWIW, I have just attempted to login again, creating a new app password for the occasion, hoping that after several months without any attempt it would work… but it instantly locked my account.

**Comment by grimmy** [comment]
So after some digging, I ended up creating a new application on developers.facebook.com, plopped the app id/secret into the source code, built installed, attempt to connect and got this better (?) error message...

> REST API is deprecated for versions v2.1 and higher (12)

So it looks like they're just blocking on the old client id, but to move to something new we have to stop using the rest API and presumably move to the graphql api? I'm just speculating there, but I don't see a way to change the API version in the developer console.

tl;dr someone is going to have to replace at least the login rest api call with something else.

**Comment by alexolog** [comment]
I asked on the Miranda-NG forum, as they seem to be unaffected.
The answer was that they use the MQTT protocol.

**Comment by grimmy** [comment]
Yeah I've started digging through their code but it's going to take some time to grok what they're doing differently.. They still have the login code that we use, but it's not in use and I haven't found the current login yet.

**Comment by truenicoco** [comment]
@grimmy I may be wrong but I wouldn't say it is 100% login-related, since every once in a while when I try, I am able to login, send and receive messages for a (short) while before triggering this 405 error.

**Comment by grimmy** [comment]
Yeah I think that's the only time the rest api is used. So once that's replaced we _might_ be good.

**Comment by grimmy** [comment]
So is anyone here currently compiling the plugin themselves? I may have a fix, but I want to get a few more people to test it before opening a pr that might not do it...

**Comment by grimmy** [comment]
well I can't reproduce this at all right now, so :shrug:

**Comment by alexolog** [comment]
No, I am on Windows and depend on the goodwill of others to provide DLLs

**Comment by grimmy** [comment]
@alexolog But if you try to connect now, do you still get the 405 error.  Like I said, the issue went away for me with no code changes.

**Comment by gmartsenkov** [comment]
@grimmy i'm willing to try it

**Comment by grimmy** [comment]
@gmartsenkov The fact that it went away without code changes show it's just a facebook issue.  Well at least it was.

**Comment by cfdourado** [comment]
I have been facing the same problem since April 4.
I have just rebuilt the src.rpm on Fedora 34, but to no avail, still the same error.
It seems that the error is now permanent, right?

**Comment by mvandemar** [comment]
I have been having this issue since last week, Centos 7, Pidgin 2.10.11-9.el7 (libpurple 2.10.11). It literally just started after working fine for years. Has any progress been made in moving towards a solution on this?

**Comment by truenicoco** [comment]
@mvandemar At first it worked a few days for me before I needed to 'verify my account', then a few hours; now using purple-facebook instantly locks my account. I wouldn't be too optimistic on this issue, but on the bright side, it may help rely less on FB messenger (it worked for me! now nobody talks to me anymore, yay!).

**Comment by mvandemar** [comment]
> So is anyone here currently compiling the plugin themselves? I may have a fix, but I want to get a few more people to test it before opening a pr that might not do it...

Since I am currently experiencing the issue I would be willing to test your potential fix.

**Comment by cfdourado** [comment]
Well, this error just stopped occurring out of the blue; I did nothing on my side and it's been working just fine since last november. Now go and figure this.... :-|

**Comment by truenicoco** [comment]
I gave my spectrum2/xmpp transport a new try these days and:

- I could login and send/receive a few messages
- After 3 days I started receiving these 405 errors again, but could unlock and still use the transport
- Now any login attempt triggers the 405 error and I cannot use it anymore.

I *may* have found out what triggers the facebook error. Examining the logs, it always seem to happen after spectrum2 sends the VCard to the facebook backend. I tried to find an option to disable updating avatar/account details but I didn't.
